Make game.

themrlolostyle opened the pull request.

Commit:
Make game.

HolyMonkey commented on Assets/Script/Ball.cs:
```diff
@@ -0,0 +1,21 @@
+
+public class Ball : MonoBehaviour
+{
+    public event UnityAction<int> ScoreChanger;
```
https://trello.com/c/CsOZfMhi/1-%D0%BD%D0%B0%D0%B7%D0%B2%D0%B0%D0%BD%D0%B8%D1%8F-%D1%81%D0%BE%D0%B1%D1%8B%D1%82%D0%B8%D0%B9-%D0%BE%D0%B1%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D1%87%D0%B8%D0%BA%D0%BE%D0%B2

HolyMonkey commented on Assets/Script/Ball.cs:
```diff
@@ -0,0 +1,21 @@
+        {
+            _score++;
+            ScoreChanger?.Invoke(_score);
+            coin.gameObject.SetActive(false);
```
Я думаю лучше если монетка сама себя отключать будет, ей нужно просто сообщить что её подобрали

HolyMonkey commented on Assets/Script/Coin.cs:
```diff
@@ -0,0 +1,12 @@
+
+public class Coin : MonoBehaviour
+{
+    [SerializeField] private float _speed;
+    private void Update()
+    {
```
Нужна пустая строка

HolyMonkey commented on Assets/Script/CoinSpawner.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public override IEnumerator Spawn()
+    {
+        while (true)
+        {
+            yield return new WaitForSeconds(_timeToSpawn);
+            for (int i = 0; i < 5; i++)
+            {
+                if (TryGetObject(out GameObject item))
+                {
+                    item.SetActive(true);
+                    item.transform.position = new Vector3(transform.position.x + i, transform.position.y, transform.position.z);
+                    item.transform.rotation = Quaternion.Euler(90, 0, 0);
+                    DisableObjectAbroadScreen();
+                }
+            }
+        }
```
Код дублируется в BarrierSpawner

HolyMonkey commented on Assets/Script/ObjectPool.cs:
```diff
@@ -0,0 +1,48 @@
+
+    private List<GameObject> _pool = new List<GameObject>();
+
+    protected void Initialize(GameObject prefab, int howMuch)
```
howMuch переименуй в count

HolyMonkey commented on Assets/Script/ObjectPool.cs:
```diff
@@ -0,0 +1,48 @@
+
+    public void DisableObjectAbroadScreen()
+    {
+        foreach (var item in _pool)
+        {
+            if (item.activeSelf == true)
+            {
+                Vector3 point = _camera.WorldToViewportPoint(item.transform.position);
+
+                if (point.x < 0)
+                {
+                    item.SetActive(false);
+                }
+            }
+        }
```
Это можно сделать дешевле по ресурсам

Commit:
Ball: change event name.
Coin: add method PickUp.
ObjectPoll: replace parametr name in Initialize, changed DIsableObjectAbroadScreen method.
Spawner: add method Spawn (not abstract).

HolyMonkey commented on Assets/Script/Ball.cs:
```diff
@@ -0,0 +1,21 @@
+        if(other.gameObject.TryGetComponent<Coin>(out Coin coin))
+        {
+            _score++;
+            ScoreChanging?.Invoke(_score);
```
Так, как это после изменения рекорда то это ScoreChanged

HolyMonkey commented on Assets/Script/Ball.cs:
```diff
@@ -0,0 +1,21 @@
+using UnityEngine;
+using UnityEngine.Events;
+
+public class Ball : MonoBehaviour
```
Так как у нас отдельно движения шара, то может и этому дадим частное название? BallWallet

HolyMonkey commented on Assets/Script/BallMove.cs:
```diff
@@ -0,0 +1,27 @@
+            _rigidbody.AddForce(Vector3.up * _powerOfJump);
+        }
+    }
+
```
Лишняя пустая строка

HolyMonkey commented on Assets/Script/BarrierSpawner.cs:
```diff
@@ -0,0 +1,7 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class BarrierSpawner : Spawner
```
Зачем?

HolyMonkey commented on Assets/Script/Coin.cs:
```diff
@@ -0,0 +1,18 @@
+
+public class Coin : MonoBehaviour
+{
+    [SerializeField] private float _speed;
```
Нужно уточнение, что это скорость вращения

HolyMonkey commented on Assets/Script/CoinSpawner.cs:
```diff
@@ -0,0 +1,7 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class CoinSpawner : Spawner
```
Зачем?

HolyMonkey commented on Assets/Script/ObjectPool.cs:
```diff
@@ -0,0 +1,39 @@
+
+    public void DisableObjectAbroadScreen()
+    {
+        foreach (var item in _pool.Where(item => item.activeSelf == true && _camera.WorldToViewportPoint(item.transform.position).x < 0))
```
Это можно сделать дешевле по ресурсам

Commit:
Fixed

HolyMonkey commented on Assets/Script/ObjectPool.cs:
```diff
@@ -0,0 +1,45 @@
+        return result != null;
+    }
+
+    public void DisableObjectAbroadScreen()
+    {
+        var activeItem = _pool.Where(item => item.activeSelf == true && PositionOnScreen(item.transform) < 0);
+        foreach (var item in activeItem)
+        {
+            item.SetActive(false);
+        }
+    }
+
+    private float PositionOnScreen(Transform transform)
+    {
+        return _camera.WorldToViewportPoint(transform.position).x;
+    }
```
Ты просто разбил код на две метода. Подумай ещё раз

Commit:
Add DeadLine

HolyMonkey commented on Assets/Script/ObjectPool.cs:
```diff
@@ -0,0 +1,31 @@
+using UnityEngine;
+using System.Linq;
+
+public class ObjectPool : MonoBehaviour
```
Нам крайне важно, чтобы ты вернул тот код с отключение объектов которые вышли за левую границу экрана но обнаружил там точку оптимизации

Commit:
Returned DisabeObcject method.

HolyMonkey commented on Assets/Script/ObjectPool.cs:
```diff
@@ -0,0 +1,42 @@
+    {
+        foreach (var item in _pool)
+        {
+            if (_camera.WorldToScreenPoint(item.transform.position).x < 0)
+            {
+                item.SetActive(false);
+            }
```
Проекция камера->мир или мир->камера 
Это очень дорого
Тут можно один раз преобразовать точку камера->мир
И потом с ней сравнивать

Commit:
ObjectPool: update DisableObjectAbroadScreen
